fix for 2024.4

Fischelsberger opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
Replacing the old "Forecast" Attribut with a service call of "weather.get_forecasts" so that it's working with 2024.4.
Tested it with Homeassistant 2024.3 and 2024.4, working with booth.

Also fixed an issue wich was within that line.
`datetime.utcfromtimestamp(fc_date / 1000)`
Some Extension provided by Homeassistant devcontainer or python itself requested that change...


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [X] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (which adds functionality to an this integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #130 #128 
- This PR is related to issue:

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] There is no commented out code in this PR.
- [ ] The code has been formatted using Black (`black --fast custom_components`)

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated to README.md

<!--
  Thank you for contributing <3
-->

Fischelsberger committed:
fix for 2024.4

golles commented on custom_components/car_wash/binary_sensor.py:
```diff
@@ -157,7 +158,15 @@ async def async_update(self):
-        forecast = wdata.attributes.get(ATTR_FORECAST)
+
+        daily_response = await self.hass.services.async_call(
+            "weather",
```
You might want to use `DOMAIN` from https://github.com/home-assistant/core/blob/dev/homeassistant/components/weather/const.py#L50

golles commented on custom_components/car_wash/binary_sensor.py:
```diff
@@ -157,7 +158,15 @@ async def async_update(self):
+
+        daily_response = await self.hass.services.async_call(
+            "weather",
+            "get_forecasts",
```
You might want to use `SERVICE_GET_FORECASTS` from https://github.com/home-assistant/core/blob/dev/homeassistant/components/weather/__init__.py#L137

golles reviewed (approved):
LGTM

Fischelsberger commented:
@golles, I've implemented your suggestions, but only tested it with 2024.4 this time.
Also im very new to the git-game, so not sure if "force push" in the old commit was the best choice...

sudoxreboot commented:
Implemented your changes. Am now getting binary result. Can't tell if it's working though, is 'Off" Wash or Don't Wash?

Fischelsberger commented:
tbh, i'm 99% sure that it always was binary.
But now as I think about, the "translation" don't make any sense with just binary results?

off = don't wash
on = wash

sudoxreboot commented:
It's working. Thank you. I believe it used to say "wash" and "don't_wash". But knowing which is which is good. Apparently it's gonna snow tomorrow, it's sunny today so I was like "why not wash?" Now I know. Thanks for the fix!

golles reviewed (approved):
@Fischelsberger thanks for addressing the comment. This looks good and I've also tested it locally and no issues!

I'm not the owner of this repo so I don't know what @Limych would require. Hopefully, he can have a look at this and also approve the usage of the workflows in this PR so we can see if CI passes.

Fischelsberger commented:
> @Fischelsberger thanks for addressing the comment. This looks good and I've also tested it locally and no issues!
> 
> I'm not the owner of this repo so I don't know what @Limych would require. Hopefully, he can have a look at this and also approve the usage of the workflows in this PR so we can see if CI passes.

@golles Wait, is this fixing your #128 then aswell?

golles commented:
> > @Fischelsberger thanks for addressing the comment. This looks good and I've also tested it locally and no issues!
> > I'm not the owner of this repo so I don't know what @Limych would require. Hopefully, he can have a look at this and also approve the usage of the workflows in this PR so we can see if CI passes.
> 
> @golles Wait, is this fixing your #128 then aswell?

Yes indeed, it's basically the same issue

Limych commented on custom_components/car_wash/binary_sensor.py:
```diff
@@ -185,7 +197,7 @@ async def async_update(self):
             if isinstance(fc_date, int):
                 fc_date = dt_util.as_local(
-                    datetime.utcfromtimestamp(fc_date / 1000)
+                    datetime.datetime.fromtimestamp(fc_date, tz=datetime.timezone.utc)
```
What are these changes for?  What problem do they solve?

Fischelsberger replied:
No Problem, but was "request" via vscode Extension probably.
Documentation: https://docs.python.org/3/library/datetime.html#datetime.datetime.utcfromtimestamp

Deprecated since version 3.12: Use [datetime.fromtimestamp()](https://docs.python.org/3/library/datetime.html#datetime.datetime.fromtimestamp) with [UTC](https://docs.python.org/3/library/datetime.html#datetime.UTC) instead.

Limych replied:
Ok. Thanks

Limych commented on custom_components/car_wash/binary_sensor.py:
```diff
@@ -157,7 +161,15 @@ async def async_update(self):
+            blocking=True,
+            return_response=True,
+        )
+        forecast = daily_response[self._weather_entity]["forecast"]
```
Please add a check for the presence of a result in the return value (daily_response).  And also add a unit test for your code.

Fischelsberger replied:
Thank you for the review and requested changes.
I've just added a new commit, with the changes which are looking good to me.

Also changed some imports, because ruff suggested these changes.

Fischelsberger committed:
fixed minor typos, added check of presence

Limych commented on custom_components/car_wash/binary_sensor.py:
```diff
@@ -157,7 +160,23 @@ async def async_update(self):
+        )
+        if not (isinstance(daily_response, dict)):
+            self._attr_is_on = None
+            raise HomeAssistantError("daily_response is empty.")
```
This error message is not clear to the user if he sees it in the log.  The user does not know what daily_response is.

Fischelsberger replied:
fixed it with new commit, by just using the original message
"Can't get forecast data! Are you sure it's the weather provider?"

by setting `forecast = None`

Limych replied:
The option is suitable, but it will be better if the error message is more unique, so that it is also more clear to developers about where the failure actually occurred.

Limych commented on custom_components/car_wash/binary_sensor.py:
```diff
@@ -157,7 +160,23 @@ async def async_update(self):
+        if not (isinstance(daily_response, dict)):
+            self._attr_is_on = None
+            raise HomeAssistantError("daily_response is empty.")
+        if (len(daily_response) == 0) or (
```
It seems to me, or is there an error in the logic of the conditional operator?

Fischelsberger replied:
oh, yeah, seems you're right!, fixed it with new commit.

Limych reviewed (changes requested):
And please add unit tests to cover code you're made.

Fischelsberger committed:
optimized Error handeling and message

Fischelsberger commented:
> And please add unit tests to cover code you're made.

Sorry @Limych, I don't really understand the whole unit test thing...
It seems to me, with the original [devcontainer.json](https://github.com/Limych/ha-car_wash/blob/dev/.devcontainer/devcontainer.json) it's no longer working.
Also the Image there [https://ghcr.io/ludeeus/devcontainer/integration:stable](https://ghcr.io/ludeeus/devcontainer/integration:stable) is Deprecated.

I've tried multiple ways, but I simply can't understand that (right now)...
I'm not a programmer myself, I just wanted to have that thing running again.
If anyone can provide me with some insights or even better, a working unit-test, I will be more then glad to implement it and run these tests and the devcontainer myself.

Limych commented:
With unit tests everything is quite simple.  We take the existing code as a basis, but we need to check all new branches according to the conditions.  It is necessary to patch self.hass.services.async_call in the test and return two variants of results - normal and error.  And check that the code reacts correctly to each option.  Tomorrow I'll try to find some time and write some sample code.

Limych commented:
Closed because these changes are no longer necessary due to the release of a new version of the component.
